# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [180]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import os
import datetime
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [181]:
df = pd.read_csv('/home/vizziwo/data/step3_output.csv', index_col=0)
df = df.reset_index(drop=True)
df.head()


Name       state  summit_elev  vertical_drop  \
0               Hilltop Ski Area      Alaska         2090            294   
1            Sunrise Park Resort     Arizona        11100           1800   
2  Yosemite Ski & Snowboard Area  California         7800            600   
3         Boreal Mountain Resort  California         7700            500   
4                    Dodge Ridge  California         8200           1600   

   trams  fastEight  fastSixes  fastQuads  quad  triple  ...  \
0      0        0.0          0          0     0       1  ...   
1      0        0.0          0          1     2       3  ...   
2      0        0.0          0          0     0       1  ...   
3      0        0.0          0          1     1       3  ...   
4      0        0.0          0          0     1       2  ...   

   SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
0               30.0       30.000000        150.000000       36.0   
1              800.0       80.000000        115.000000       49.0   
2               88.0      174.873239        110.000000       84.0   
3              380.0      200.000000        150.000000       54.0   
4              862.0      174.873239        115.103943       69.0   

   averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
0             69.0          30.0      34.00000              152.0   
1            250.0          74.0      78.00000              104.0   
2            300.0          47.0      47.00000              107.0   
3            400.0          49.0      64.16681              150.0   
4            350.0          78.0      78.00000              140.0   

   NightSkiing_ac  clusters  
0            30.0         2  
1            80.0         0  
2             0.0         0  
3           200.0         0  
4             0.0         0  

[5 rows x 26 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [182]:
state=df['state']
df = pd.concat([df.drop('state',axis=1), pd.get_dummies(state)], axis=1)


In [183]:
df.head()

Name  summit_elev  vertical_drop  trams  \
0               Hilltop Ski Area         2090            294      0   
1            Sunrise Park Resort        11100           1800      0   
2  Yosemite Ski & Snowboard Area         7800            600      0   
3         Boreal Mountain Resort         7700            500      0   
4                    Dodge Ridge         8200           1600      0   

   fastEight  fastSixes  fastQuads  quad  triple  double  ...  Rhode Island  \
0        0.0          0          0     0       1       0  ...             0   
1        0.0          0          1     2       3       1  ...             0   
2        0.0          0          0     0       1       3  ...             0   
3        0.0          0          1     1       3       1  ...             0   
4        0.0          0          0     1       2       5  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 60 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [184]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X = df.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = df.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler()

scaler.fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [185]:
X_scaled, y

(array([[-0.48159014, -0.93919067,  0.        , ..., -0.13323468,
         -0.248452  , -0.17303214],
        [ 2.32208485,  1.64447388,  0.        , ..., -0.13323468,
         -0.248452  , -0.17303214],
        [ 1.29521166, -0.41422297,  0.        , ..., -0.13323468,
         -0.248452  , -0.17303214],
        ...,
        [ 1.29770105,  1.2516059 ,  0.        , ..., -0.13323468,
         -0.248452  ,  5.77927331],
        [ 1.87492825,  0.2548535 ,  0.        , ..., -0.13323468,
         -0.248452  ,  5.77927331],
        [ 1.82420694,  0.44356738,  0.        , ..., -0.13323468,
         -0.248452  ,  5.77927331]]), 0      34.00000
 1      78.00000
 2      47.00000
 3      64.16681
 4      78.00000
          ...   
 167    48.00000
 168    42.00000
 169    59.00000
 170    49.00000
 171    49.00000
 Name: AdultWeekend, Length: 172, dtype: float64)

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [186]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train,X_test,y_train,y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [187]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [188]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)

## Review Model Outcomes — Iterate over additional models as needed

In [189]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
# To do so, you will need to import them from sklearn.metrics. 
# You can plug y_test and y_pred into the functions to evaluate the model

In [190]:
explained_variance_score(y_test, y_pred)

-2.6146896892416963e+20

In [191]:
mean_absolute_error(y_test, y_pred)

61513366881.14965

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [192]:
model.intercept_

-15378341661.542433

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [193]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)


Coefficient
total_chairs       2.570839e+13
fastEight          1.376312e+13
double             1.366951e+13
surface            1.342053e+13
triple             1.145968e+13
trams              9.016867e+12
quad               6.535489e+12
fastQuads          5.464577e+12
fastSixes          1.797909e+12
New York           4.236764e+11
Michigan           3.729005e+11
California         3.367039e+11
New Hampshire      3.233753e+11
Pennsylvania       3.233753e+11
Wisconsin          3.092823e+11
Vermont            2.783328e+11
Minnesota          2.783328e+11
New Mexico         2.425086e+11
Idaho              2.425086e+11
Maine              2.220449e+11
Connecticut        2.220449e+11
Wyoming            2.220449e+11
Massachusetts      1.991967e+11
Colorado           1.991967e+11
Ohio               1.991967e+11
North Carolina     1.730221e+11
West Virginia      1.730221e+11
Utah               1.730221e+11
Washington         1.730221e+11
Oregon             1.730221e+11
Iowa               1.730221e+11
Montana            1.730221e+11
Illinois           1.730221e+11
Virginia           1.416893e+11
Missouri           1.416893e+11
Indiana            1.416893e+11
Tennessee          1.004837e+11
New Jersey         1.004837e+11
Rhode Island       1.004837e+11
Arizona            1.004837e+11
South Dakota       1.004837e+11
Alaska             1.004837e+11
Maryland           1.344462e+08
Nevada             3.909329e+06
AdultWeekday       8.207031e+00
clusters           2.170410e+00
summit_elev        1.741440e+00
vertical_drop      1.387173e+00
averageSnowfall    1.348633e+00
projectedDaysOpen  1.303223e+00
Runs               1.261719e+00
daysOpenLastYear   1.099609e+00
LongestRun_mi      1.075684e+00
TerrainParks       9.931641e-01
NightSkiing_ac     7.373047e-01
Snow Making_ac     2.978516e-01
SkiableTerrain_ac  2.568359e-01
yearsOpen          2.265625e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [194]:
df2 = df.loc[:,'summit_elev':'clusters'].drop(columns='AdultWeekend')
y = df.AdultWeekend
scaler2 = preprocessing.StandardScaler()
scaler2.fit(df2)
df2_scaled = scaler2.transform(df2)

In [195]:
y = y.ravel()
df2_train, df2_test, y_train, y_test = train_test_split(df2_scaled, y, test_size=0.25, random_state=1)

In [196]:
lm2 = linear_model.LinearRegression()
model2 = lm2.fit(df2_train,y_train)

In [197]:
y_pred = model2.predict(df2_test)

In [198]:
print(explained_variance_score(y_test, y_pred), mean_absolute_error(y_test, y_pred))

0.8175746547910714 6.187862966510734


In [199]:
model2.intercept_

57.39147073323086

In [200]:
pd.DataFrame(abs(lm2.coef_), df2.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)


Coefficient
AdultWeekday       9.690891e+00
summit_elev        3.474329e+00
vertical_drop      3.468729e+00
daysOpenLastYear   2.163140e+00
LongestRun_mi      1.918399e+00
SkiableTerrain_ac  1.800170e+00
TerrainParks       1.759516e+00
quad               1.437082e+00
projectedDaysOpen  1.300101e+00
averageSnowfall    1.043342e+00
surface            1.039398e+00
double             1.008357e+00
clusters           9.459791e-01
triple             7.070347e-01
fastQuads          6.506041e-01
Runs               5.799245e-01
total_chairs       5.357639e-01
yearsOpen          2.989631e-01
Snow Making_ac     1.373068e-01
NightSkiing_ac     3.616495e-02
fastSixes          8.881784e-16
fastEight          2.220446e-16
trams              0.000000e+00

**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [202]:
df.columns

Index(['Name', 'summit_elev', 'vertical_drop', 'trams', 'fastEight',
       'fastSixes', 'fastQuads', 'quad', 'triple', 'double', 'surface',
       'total_chairs', 'Runs', 'TerrainParks', 'LongestRun_mi',
       'SkiableTerrain_ac', 'Snow Making_ac', 'daysOpenLastYear', 'yearsOpen',
       'averageSnowfall', 'AdultWeekday', 'AdultWeekend', 'projectedDaysOpen',
       'NightSkiing_ac', 'clusters', 'Alaska', 'Arizona', 'California',
       'Colorado', 'Connecticut', 'Idaho', 'Illinois', 'Indiana', 'Iowa',
       'Maine', 'Maryland', 'Massachusetts', 'Michigan', 'Minnesota',
       'Missouri', 'Montana', 'Nevada', 'New Hampshire', 'New Jersey',
       'New Mexico', 'New York', 'North Carolina', 'Ohio', 'Oregon',
       'Pennsylvania', 'Rhode Island', 'South Dakota', 'Tennessee', 'Utah',
       'Vermont', 'Virginia', 'Washington', 'West Virginia', 'Wisconsin',
       'Wyoming'],
      dtype='object')

In [203]:
X3 = df.loc[: , 'trams':'clusters'].drop(columns=['AdultWeekend'])

In [ ]:
#BASE_ELEV HAS ALREADY BEEN DROPPED BECAUSE ITS CORRELACTION WAS OVER .95 IN STEP 3!!!

In [204]:
y = df.AdultWeekend
scaler3 = preprocessing.StandardScaler()
scaler3.fit(X3)
X3_scaled = scaler3.transform(X3)

In [205]:
y = y.ravel()

In [206]:
X3_train,X3_test,y_train,y_test = train_test_split(X3_scaled, y, test_size=0.25, random_state = 1)

In [207]:
lm3 = linear_model.LinearRegression()
model3 = lm3.fit(X3_train, y_train)

In [208]:
y_pred = model3.predict(X3_test)

In [209]:
explained_variance_score(y_test, y_pred), mean_absolute_error(y_test, y_pred)

(0.8239888393384389, 6.162988281762991)

In [211]:
model3.intercept_

57.549194969754886

In [218]:
pd.DataFrame(abs(model3.coef_), X3.columns, columns=['Coefficient']).sort_values(by='Coefficient', ascending=False)

Coefficient
AdultWeekday       1.027962e+01
LongestRun_mi      2.405090e+00
daysOpenLastYear   2.316623e+00
clusters           1.623535e+00
Runs               1.520599e+00
projectedDaysOpen  1.496139e+00
TerrainParks       1.470295e+00
surface            1.334153e+00
quad               1.333028e+00
SkiableTerrain_ac  1.324274e+00
double             9.957783e-01
averageSnowfall    7.753856e-01
triple             4.339187e-01
total_chairs       3.915990e-01
yearsOpen          2.617310e-01
fastQuads          1.237159e-01
Snow Making_ac     9.596115e-02
NightSkiing_ac     1.352757e-02
fastSixes          4.440892e-16
fastEight          0.000000e+00
trams              0.000000e+00

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -2.61468 | 61513366881.14965 |-|
| Model 2. | 0.81757 | 6.18786 |'state'|
| Model 3. | 0.82398 | 6.16298 |'state','summit_elev','base_elev'|

Model Selection:

# Model 3 is way under the limit of '40' for Mean Abs Error.
# It's also closer to '1' than the other models in R-Squared
# Model 3 is the best

In [ ]:
|